cc/utils/cose/cose.cc

```diff
@@ -0,0 +1,127 @@
+/*
+ * Copyright 2024 The Project Oak Authors
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "cc/utils/cose/cose.h"
+
+#include <memory>
+#include <string>
+#include <vector>
+
+#include "absl/status/statusor.h"
+#include "absl/strings/string_view.h"
+#include "libcppbor/include/cppbor/cppbor.h"
+#include "libcppbor/include/cppbor/cppbor_parse.h"
+
+namespace oak::utils::cose {
+
+absl::StatusOr<CoseSign1> CoseSign1::Deserialize(const std::vector<uint8_t>& data) {
+  auto [item, end, error] = cppbor::parse(data);
+  if (!error.empty()) {
+    return absl::InvalidArgumentError(absl::StrCat("couldn't parse COSE_Sign1: ", error));
+  }
+  if (item->type() != cppbor::ARRAY) {
+    return UnexpectedCborTypeError("COSE_Sign1", cppbor::ARRAY, item->type());
+  }
+  const cppbor::Array* array = item->asArray();
+  if (array->size() != 4) {
+    return absl::InvalidArgumentError(
+        absl::StrCat("invalid COSE_Sign1 CBOR array size, expected 4, found ", array->size()));
+  }
+
+  auto& protected_headers = array->get(0);
+  if (protected_headers->type() != cppbor::BSTR) {
+    return UnexpectedCborTypeError("protected_headers", cppbor::BSTR, protected_headers->type());
+  }
+  auto& unprotected_headers = array->get(1);
+  if (unprotected_headers->type() != cppbor::MAP) {
+    return UnexpectedCborTypeError("unprotected_headers", cppbor::MAP, unprotected_headers->type());
+  }
+  auto& payload = array->get(2);
+  if (payload->type() != cppbor::BSTR) {
+    return UnexpectedCborTypeError("payload", cppbor::BSTR, payload->type());
+  }
+  auto& signature = array->get(3);
+  if (signature->type() != cppbor::BSTR) {
+    return UnexpectedCborTypeError("signature", cppbor::BSTR, signature->type());
+  }
+
+  return CoseSign1(protected_headers->asBstr(), unprotected_headers->asMap(), payload->asBstr(),
+                   signature->asBstr(), std::move(item));
+}
+
+absl::StatusOr<CoseKey> CoseKey::Deserialize(const std::vector<uint8_t>& data) {
```

Comment:
nit: perhaps clarify that this is only for encryption keys (e.g. name it as DeserializeHpkePublicKey) since this won't work with all COSE keys.

Reply:
Done